common/util/tree_operations.h

```diff
@@ -0,0 +1,926 @@
+// Copyright 2017-2022 The Verible Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This file contains generic functions and trait classes for operating on tree
+// structures.
+//
+// TreeNode concept
+// ----------------
+//
+// Class `T` must contain at least the following member in order to fulfill the
+// TreeNode concept:
+//
+// - `Container<T>& Children()`:
+//   Returns reference to a STL-like container or a range containing the node's
+//   children.
+//   The Container must at least support `begin()` and `end()` methods.
+//   Pointer and iterator stability is not required.
+//
+// To check whether class `T` fulfills the concept one can use the
+// `TreeNodeTraits<T>` traits struct. It is defined only for classes fulfilling
+// the TreeNode concept, what makes it suitable for SFINAE tests. For
+// readability and consistency with Value and Parent checks (described below)
+// there is a member `TreeNodeTraits<T>::available` which is always true.
+//
+// Optional members:
+//
+// - `T* Parent()`:
+//   Returns pointer to a parent node or nullptr when the node is a tree root.
+//   Types with this member can be detected by checking whether
+//   `TreeNodeTraits<T>::Parent::available` is true.
+//
+// - `ValueType& Value()`:
+//   Returns reference to a value stored in the node.
+//   Types with this member can be detected by checking whether
+//   `TreeNodeTraits<T>::Parent::available` is true.
+//
+// - `subnodes_type` (typename):
+//   Container type which should be used for storing arrays of detached
```

Comment:
// Contiguous storage (array) is required (?) or only bidirectional iterability?

Reply:
"Array" is a misleading word here - replaced with "collections". I've also added more explanation. New comment:
```
// - `subnodes_type` (typename):
//   Container type which should be used for storing collections of detached
//   child nodes.
//   It must be move-assignable to, and move-constructible from a value returned
//   by `Children()` method. It must provide at least the same STL container's
//   interface and iterator/pointer stability guarantees as a return type of
//   `Children()` method.
//   The use case where this alias is useful is when `Children()` returns a
//   container wrapped by some internal wrapper type that is either not
//   constructible outside of the node class or has some extra logic that is
//   not useful when used outside of a node. The `subnodes_type` in such case
//   should be an alias to the container's real type.
//   When the type is absent, it is assumed to be the type returned by
//   `Children()` method with reference and other modifiers removed.
//    It is available through `TreeNodeTraits<T>::Children::container_type`.
```